allennlp/models/coreference_resolution/coref_v2.py

```diff
@@ -0,0 +1,744 @@
+            A tensor of shape (batch_size, num_spans), representing the cluster ids
+            of each span, or -1 for those which do not appear in any clusters.
+
+        Returns
+        -------
+        An output dictionary consisting of:
+        top_spans : ``torch.IntTensor``
+            A tensor of shape ``(batch_size, num_spans_to_keep, 2)`` representing
+            the start and end word indices of the top spans that survived the pruning stage.
+        antecedent_indices : ``torch.IntTensor``
+            A tensor of shape ``(num_spans_to_keep, max_antecedents)`` representing for each top span
+            the index (with respect to top_spans) of the possible antecedents the model considered.
+        predicted_antecedents : ``torch.IntTensor``
+            A tensor of shape ``(batch_size, num_spans_to_keep)`` representing, for each top span, the
+            index (with respect to antecedent_indices) of the most likely antecedent. -1 means there
+            was no predicted link.
+        loss : ``torch.FloatTensor``, optional
+            A scalar loss to be optimised.
+        """
+        # Shape: (batch_size, document_length, embedding_size)
+        text_embeddings = self._lexical_dropout(self._text_field_embedder(text))
+
+        document_length = text_embeddings.size(1)
+        num_spans = span_starts.size(1)
+
+        # Shape: (batch_size, document_length)
+        text_mask = util.get_text_field_mask(text).float()
+
+        # Shape: (batch_size, num_spans, 1)
+        span_mask = (span_starts >= 0).float()
+        # IndexFields return -1 when they are used as padding. As we do
+        # some comparisons based on span widths when we attend over the
+        # span representations that we generate from these indices, we
+        # need them to be <= 0. This is only relevant in edge cases where
+        # the number of spans we consider after the pruning stage is >= the
+        # total number of spans, because in this case, it is possible we might
+        # consider a masked span.
+        span_starts = F.relu(span_starts.float()).long()
+        span_ends = F.relu(span_ends.float()).long()
+
+        # Shape: (batch_size, num_spans, embedding_size)
+        span_embeddings = self._compute_span_representations(text_embeddings,
+                                                             text_mask,
+                                                             span_starts,
+                                                             span_ends)
+        # Prune based on mention scores.
+        num_spans_to_keep = int(math.floor(self._spans_per_word * document_length))
+
+        (top_span_embeddings, top_span_mask,
+         top_span_indices, top_span_mention_scores) = self._mention_pruner(span_embeddings,
+                                                                           span_mask.squeeze(-1),
+                                                                           num_spans_to_keep)
```

Comment:
Here is where the `SpanPruner` is added to the coref model. It replaces the [`_prune_and_sort_spans`](https://github.com/allenai/allennlp/blob/master/allennlp/models/coreference_resolution/coref.py#L531) method.